File: Day24/Python/Main.py
import os

class Hexagon:
    coordLookup = {}

    def __init__(self, x, y):
        self.black = False
        self.x = x
        self.y = y
        self.coordLookup[(x, y)] = self;
        self.east = self.southEast = self.southWest = self.west = self.northWest = self.northEast = None
        self.setEast(self.coordLookup.get((x + 1, y), None))
        self.setSouthEast(self.coordLookup.get((x + 0.5, y + 1), None))
        self.setSouthWest(self.coordLookup.get((x - 0.5, y + 1), None))
        self.setWest(self.coordLookup.get((x - 1, y), None))
        self.setNorthWest(self.coordLookup.get((x - 0.5, y - 1), None))
        self.setNorthEast(self.coordLookup.get((x + 0.5, y - 1), None))

    def toggle(self):
        self.black = not self.black
        return self.black

    def hasEast(self):
        return self.east != None

    def getEast(self):
        return self.east

    def getOrCreateEast(self):
        if not self.hasEast():
            self.setEast(Hexagon(self.x + 1, self.y))

        return self.getEast()

    def setEast(self, east):
        if east == None:
            return

        self.east = east

        if not east.hasWest():
            east.setWest(self)

    def hasSouthEast(self):
        return self.southEast != None

    def getSouthEast(self):
        return self.southEast

    def getOrCreateSouthEast(self):
        if not self.hasSouthEast():
            self.setSouthEast(Hexagon(self.x + 0.5, self.y + 1))

        return self.getSouthEast()

    def setSouthEast(self, southEast):
        if southEast == None:
            return

        self.southEast = southEast

        if not southEast.hasNorthWest():
            southEast.setNorthWest(self)

    def hasSouthWest(self):
        return self.southWest != None

    def getSouthWest(self):
        return self.southWest

    def getOrCreateSouthWest(self):
        if not self.hasSouthWest():
            self.setSouthWest(Hexagon(self.x - 0.5, self.y + 1))

        return self.getSouthWest()

    def setSouthWest(self, southWest):
        if southWest == None:
            return

        self.southWest = southWest;

        if not southWest.hasNorthEast():
            southWest.setNorthEast(self)

    def hasWest(self):
        return self.west != None

    def getWest(self):
        return self.west

    def getOrCreateWest(self):
        if not self.hasWest():
            self.setWest(Hexagon(self.x - 1, self.y))

        return self.getWest()

    def setWest(self, west):
        if west == None:
            return

        self.west = west

        if not west.hasEast():
            west.setEast(self)

    def hasNorthWest(self):
        return self.northWest != None

    def getNorthWest(self):
        return self.northWest

    def getOrCreateNorthWest(self):
        if not self.hasNorthWest():
            self.setNorthWest(Hexagon(self.x - 0.5, self.y - 1))

        return self.getNorthWest()

    def setNorthWest(self, northWest):
        if northWest == None:
            return

        self.northWest = northWest

        if not northWest.hasSouthEast():
            northWest.setSouthEast(self)

    def hasNorthEast(self):
        return self.northEast != None

    def getNorthEast(self):
        return self.northEast

    def getOrCreateNorthEast(self):
        if not self.hasNorthEast():
            self.setNorthEast(Hexagon(self.x + 0.5, self.y - 1))

        return self.getNorthEast()

    def setNorthEast(self, northEast):
        if northEast == None:
            return

        self.northEast = northEast

        if not northEast.hasSouthWest():
            northEast.setSouthWest(self)

def main():
    input = os.path.dirname(os.getcwd())
    input = os.path.join(input, "input")
    if not os.path.exists(input):
        print(input + " does not exist!")
        input = os.path.join(os.path.dirname(input), "input.txt")
        print(f"trying {input} instead.")
        if not os.path.exists(input):
            print(input + " does not exist!")
            return;

    center = Hexagon(0, 0)
    black = 0
    for line in open(input, 'r').readlines():
        current = center
        i = 0
        while i < len(line):
            c = line[i]
            if c == 'e':
                current = current.getOrCreateEast()
            elif c == 's':
                i += 1
                if line[i] == 'e':
                    current = current.getOrCreateSouthEast()
                else:
                    current = current.getOrCreateSouthWest()
            elif c == 'w':
                current = current.getOrCreateWest()
            elif c == 'n':
                i += 1
                if line[i] == 'w':
                    current = current.getOrCreateNorthWest()
                else:
                    current = current.getOrCreateNorthEast()

            i += 1
        if current.toggle():
            black += 1
        else:
            black -= 1

    print(f"Part 1 result is {black}.")

if __name__ == '__main__':
    main()

File: Day24/Java/Main.java
import java.io.File;
import java.io.IOException;
import java.io.StringReader;
import java.nio.file.Files;
import java.util.HashMap;
import java.util.Map;
import java.util.Objects;

public class Main {

	public static void main(String[] args) throws IOException {
		File input = new File("input").getAbsoluteFile();
		if (!input.exists()) {
			System.out.println(input.getPath() + " does not exist!");
			System.out.println("trying " + input.getPath() + ".txt instead.");
			input = new File(input.getPath() + ".txt");
			if (!input.exists()) {
				System.err.println("neither " + input.getPath().substring(0, input.getPath().length() - 4) + ",");
				System.err.println("nor " + input.getPath() + " exists!");
				return;
			}
		}

		Hexagon center = new Hexagon(0, 0);
		int black = 0;
		for (String line : Files.readAllLines(input.toPath())) {
			StringReader reader = new StringReader(line);
			Hexagon current = center;
			ready: while (reader.ready()) {
				switch (reader.read()) {
				case 'e':
					current = current.getOrCreateEast();
					break;
				case 's':
					if (reader.read() == 'e') {
						current = current.getOrCreateSouthEast();
					} else {
						current = current.getOrCreateSouthWest();
					}
					break;
				case 'w':
					current = current.getOrCreateWest();
					break;
				case 'n':
					if (reader.read() == 'w') {
						current = current.getOrCreateNorthWest();
					} else {
						current = current.getOrCreateNorthEast();
					}
					break;
				case -1:
					break ready;
				}
			}
			black += current.toggle() ? 1 : -1;
		}

		System.out.println("The part 1 result is " + black + ".");
	}

	private static class Pos {
		public float x, y;

		public Pos(float x, float y) {
			this.x = x;
			this.y = y;
		}

		@Override
		public int hashCode() {
			return Objects.hash(x, y);
		}

		@Override
		public boolean equals(Object obj) {
			if (this == obj)
				return true;
			if (obj == null)
				return false;
			if (getClass() != obj.getClass())
				return false;
			Pos other = (Pos) obj;
			return Float.floatToIntBits(x) == Float.floatToIntBits(other.x)
					&& Float.floatToIntBits(y) == Float.floatToIntBits(other.y);
		}
	}

	private static class Hexagon {
		private static final Map<Pos, Hexagon> COORD_LOOKUP = new HashMap<>();

		private Hexagon east;
		private Hexagon southEast;
		private Hexagon southWest;
		private Hexagon west;
		private Hexagon northWest;
		private Hexagon northEast;
		private float x, y;
		private boolean black;

		public Hexagon(float x, float y) {
			COORD_LOOKUP.put(new Pos(this.x = x, this.y = y), this);
			setEast(COORD_LOOKUP.get(new Pos(x + 1, y)));
			setSouthEast(COORD_LOOKUP.get(new Pos(x + 0.5f, y + 1)));
			setSouthWest(COORD_LOOKUP.get(new Pos(x - 0.5f, y + 1)));
			setWest(COORD_LOOKUP.get(new Pos(x - 1, y)));
			setNorthWest(COORD_LOOKUP.get(new Pos(x - 0.5f, y - 1)));
			setNorthEast(COORD_LOOKUP.get(new Pos(x + 0.5f, y - 1)));
		}

		public boolean toggle() {
			return black = !black;
		}

		public boolean hasEast() {
			return east != null;
		}

		public Hexagon getEast() {
			return east;
		}

		public Hexagon getOrCreateEast() {
			if (!hasEast()) {
				setEast(new Hexagon(x + 1, y));
			}

			return getEast();
		}

		public void setEast(Hexagon east) {
			if (east == null) {
				return;
			}

			this.east = east;

			if (!east.hasWest()) {
				east.setWest(this);
			}
		}

		public boolean hasSouthEast() {
			return southEast != null;
		}

		public Hexagon getSouthEast() {
			return southEast;
		}

		public Hexagon getOrCreateSouthEast() {
			if (!hasSouthEast()) {
				setSouthEast(new Hexagon(x + 0.5f, y + 1));
			}

			return getSouthEast();
		}

		public void setSouthEast(Hexagon southEast) {
			if (southEast == null) {
				return;
			}

			this.southEast = southEast;

			if (!southEast.hasNorthWest()) {
				southEast.setNorthWest(this);
			}
		}

		public boolean hasSouthWest() {
			return southWest != null;
		}

		public Hexagon getSouthWest() {
			return southWest;
		}

		public Hexagon getOrCreateSouthWest() {
			if (!hasSouthWest()) {
				setSouthWest(new Hexagon(x - 0.5f, y + 1));
			}

			return getSouthWest();
		}

		public void setSouthWest(Hexagon southWest) {
			if (southWest == null) {
				return;
			}

			this.southWest = southWest;

			if (!southWest.hasNorthEast()) {
				southWest.setNorthEast(this);
			}
		}

		public boolean hasWest() {
			return west != null;
		}

		public Hexagon getWest() {
			return west;
		}

		public Hexagon getOrCreateWest() {
			if (!hasWest()) {
				setWest(new Hexagon(x - 1, y));
			}

			return getWest();
		}

		public void setWest(Hexagon west) {
			if (west == null) {
				return;
			}

			this.west = west;

			if (!west.hasEast()) {
				west.setEast(this);
			}
		}

		public boolean hasNorthWest() {
			return northWest != null;
		}

		public Hexagon getNorthWest() {
			return northWest;
		}

		public Hexagon getOrCreateNorthWest() {
			if (!hasNorthWest()) {
				setNorthWest(new Hexagon(x - 0.5f, y - 1));
			}

			return getNorthWest();
		}

		public void setNorthWest(Hexagon northWest) {
			if (northWest == null) {
				return;
			}

			this.northWest = northWest;

			if (!northWest.hasSouthEast()) {
				northWest.setSouthEast(this);
			}
		}

		public boolean hasNorthEast() {
			return northEast != null;
		}

		public Hexagon getNorthEast() {
			return northEast;
		}

		public Hexagon getOrCreateNorthEast() {
			if (!hasNorthEast()) {
				setNorthEast(new Hexagon(x + 0.5f, y - 1));
			}

			return getNorthEast();
		}

		public void setNorthEast(Hexagon northEast) {
			if (northEast == null) {
				return;
			}

			this.northEast = northEast;

			if (!northEast.hasSouthWest()) {
				northEast.setSouthWest(this);
			}
		}
	}
}
